packages/aws-cdk/lib/commands/init/init.ts

```diff
@@ -76,8 +71,116 @@ export async function cliInit(options: CliInitOptions) {
+  try {
+    const template = await InitTemplate.fromPath(templatePath);
+
+    if (template.languages.length === 0) {
+      throw new ToolkitError('Custom template must contain at least one language directory');
+    }
+
+    return template;
+  } catch (e: any) {
+    throw new ToolkitError(`Failed to load template from path: ${templatePath}. ${e.message}`);
+  }
+}
+
+/**
+ * Load a built-in template by name
+ * @param ioHelper - IO helper for user interaction
+ * @param type - Template type name
+ * @param language - Programming language filter
+ * @returns Promise resolving to the loaded InitTemplate
+ */
+async function loadBuiltinTemplate(ioHelper: IoHelper, type?: string, language?: string): Promise<InitTemplate> {
+  if (!type && !language) {
+    await printAvailableTemplates(ioHelper, language);
+    throw new ToolkitError('No template specified. Please specify a template name.');
+  }
+
+  if (!type) {
+    await printAvailableTemplates(ioHelper, language);
+    throw new ToolkitError('No template specified. Please specify a template name.');
+  }
+
+  const template = (await availableInitTemplates()).find((t) => t.hasName(type));
+
+  if (!template) {
+    await printAvailableTemplates(ioHelper, language);
+    throw new ToolkitError(`Unknown init template: ${type}`);
+  }
+
+  return template;
+}
+
+/**
+ * Resolve the programming language for the template
+ * @param ioHelper - IO helper for user interaction
+ * @param template - The template to resolve language for
+ * @param requestedLanguage - User-requested language (optional)
+ * @returns Promise resolving to the selected language
+ */
+async function resolveLanguage(ioHelper: IoHelper, template: InitTemplate, requestedLanguage?: string): Promise<string> {
+  let language = requestedLanguage;
+
+  // Auto-detect language for single-language templates
+  if (!language && template.languages.length === 1) {
+    language = template.languages[0];
+    await ioHelper.defaults.info(
+      `No --language was provided, but '${template.name}' supports only '${language}', so defaulting to --language=${language}`,
+    );
+  }
+
+  if (!language) {
+    await ioHelper.defaults.info(`Available languages for ${chalk.green(template.name)}: ${template.languages.map((l) => chalk.blue(l)).join(', ')}`);
+    throw new ToolkitError('No language was selected');
+  }
+
+  return language;
+}
+
+/**
+ * Get valid CDK language directories from a template path
+ * @param templatePath - Path to the template directory
+ * @returns Promise resolving to array of supported language names
+ */
+async function getLanguageDirectories(templatePath: string): Promise<string[]> {
+  const result: string[] = [];
+  const supportedLanguages = ['typescript', 'javascript', 'python', 'java', 'csharp', 'fsharp', 'go'];
+
+  try {
+    const entries = await fs.readdir(templatePath, { withFileTypes: true });
+
+    for (const entry of entries) {
```

Comment:
This is synchronous, blocking code. We can improve here by promisifying it with something like Promise.all(), which will parallelize expensive calls that we are awaiting inside loops.